### ISSUE TITLE
Implement contains(...) filter predicate (RFC 254)

### ISSUE THREAD

**Issue by rvantonder** [sourcegraph/sourcegraph#18584]
See [RFC 254](https://docs.google.com/document/d/1_m63fsBMAtqaq3GA_aMzKUPxD3yxTy8d12lJE6qN6PU/edit#heading=h.5o1uqjuq3yj5) for full context. This issue tracks the implementation plan for the `contains` filter predicate in the search query.

We should implement `contains` for the `repo` and `file` filters. We'll do it for `repo` first, then expand to `file` time allowing. Later, we may want to generalize `contains` to other filters like commits, but it is not in scope of this plan nor RFC 254.

The backend will need quite a bit of work, since we want to essentially perform subqueries based on `contains` to extract values that then constrain and scope the parent query. Realistically, I think we may only get to `repo` within the time allocated, and enable the queries and behavior as documented in [RFC 254](https://docs.google.com/document/d/1_m63fsBMAtqaq3GA_aMzKUPxD3yxTy8d12lJE6qN6PU/edit#heading=h.5o1uqjuq3yj5), and `file` is a stretch goal.

**Backend**

- [x] Add scaffolding to recognize filter predicates. This should be a map so we can do a lookup on a filter predicate and then dispatch to a function that does the work. We will only add `contains` as the first filter predicate in this map.

- [x] Parse filter predicate values. There are a couple of ways to do this. My thinking is to add a post-processing step after parsing a query into it's canonical form (simple `field:value`) and then parsing out the value and perform the behavior (i.e., filter predicates are "plugin" logic after parsing a query, and we activate the plugin logic based on recognizing a registered predicate).

- [x] Implement `contains` logic for repo. The reference of behavior we want to implement [is documented in RFC 254](https://docs.google.com/document/d/1_m63fsBMAtqaq3GA_aMzKUPxD3yxTy8d12lJE6qN6PU/edit#heading=h.5o1uqjuq3yj5). 
  - [x] Syntax: We will start with simple `field:value` inputs for `contains`, where `field` can be `file` or `content` and values are interpreted as regular expressions. Am a bit handy wavy on the details here, but we can reuse the field/value parser/scanners for the arguments of `value` to extract these values.
  - [x] Semantics: `repo:contains(v)` essentially means "substitute for `contains` the repository names that satisfy the query `v`". The idea is that we want to substitute `contains(v)` for something like `repo:foo|bar|baz` so that we can simply evaluate the query with our existing search engine. Important for the implementation, is the order in which we evaluate `repo:contains(...)` predicates. If we have a query like `repo:sourcegraph repo:contains(file:.*)`, we could in theory evaluate `contains(...)` first (resolving to all repos), and then intersect with the values satisfying `repo:sourcegraph`. This would be inefficient. More efficient would be if we resolved `contains` with respect to existing `repo:sourcegraph`, i.e., we run the search `repo:sourcegraph file:.* select:repo` and substitute the output of this value for `repo:...` and run our query. Thus, `repo:contains(...)` should be implemented as a search query using `select`, and give precedence to other `repo:` values in the query before running `contains(...)`. 
  - [x] Proper predicate parsing #19075 

- [ ] Deprecate `repoHasFile` depends on #20337

Additional scope beyond RFC 254
  - [x] Support `repo:contains(file:foo content:bar)`
  - [ ] Support `repo:contains(-file:foo -content:bar)`
  - [ ] Negation on `repo`, e.g., `-repo:cointains(...)`
  - [ ] Add support for `repo:contains(x) repo:contains(y)` i.e., `repo:contains(x) and repo:contains(y)`, e.g., make this work [`repo:^github\.com/sourcegraph/ repo:contains(file:README.md) repo:contains(file:CHANGELOG)`](https://sourcegraph.com/search?q=repo:%5Egithub%5C.com/sourcegraph/+repo:contains%28file:README.md%29+repo:contains%28file:CHANGELOG%29&patternType=regexp). At this point, we've fulfilled the scope of RFC 254 and deprecated `repoHasFile`. The following are stretch goals.
  
  Examples:
    - `repo:foo repo:contains(file:bar)` implies `repo:(repo:foo repo:contains(file:bar) select:repo)`. The latter query can be expressed in Zoekt as `(type:repo repo: file:bar)`, but that's not possible for us right now--we can rely on our `select` implementation. The result of `repo:foo repo:contains(file:bar) select:repo` would be transformed into a regular expression value that we then replace the `repo:` value in the original query and run this search. We have some flexibility on figuring out how to evaluate subexpressions, i.e., if the regular expression is very large (many repos satisfy `contains`) then we can optimize by splitting it into `or` queries for evaluation, for example. For simplicity, it's easiest to assume we'll build a regular expression for a start.
    - `repogroup:foo -repo:bar repo:contains(content:qux)` implies `repo:(repogroup:foo -repo:bar content:qux select:repo)`
    - Shelved the below, cf. https://github.com/sourcegraph/sourcegraph/pull/19049#discussion_r592661271
    - ~`repo:foo -file:bar file:contains(content:baz)` implies `repo:foo file:(-file:bar content:baz select:file)`. I.e., for `file:contains(...)` we'll do similarly.~
    - ~Complex case (we won't implement this yet, but I want to demonstrate how it pans out): `repo:foo repo:contains(file:bar) -file:baz file:contains(content:qux) hey` implies ~
    ```
    repo:(repo:foo file:(file:bar -file:baz content:qux select:file) select:repo) 
    file:(file:bar -file:baz content:qux select:file) 
    hey
    ```
    
    Shelving interaction of `repo:contains` and `file:contains` https://github.com/sourcegraph/sourcegraph/pull/19049#discussion_r592661271 ~I.e., For combinations of `repo:contains` and `file:contains`, we'll need to do subquery expansion. We will punt on this for RFC purposes right now. If we implement things nicely, it should be straightforward to evaluate this--it would only mean getting the query in the right shape--we are just substituting the expansion of `file:` inside the `repo:` routine, for example.~

Extra notes: it'd be worthwhile to look at how `repoHasFile` is currently implemented. I don't think we can reuse much of this code, but it would be interesting if we can learn anything about how it optimizes this use case, and maybe we can borrow parts of the approach.

**Frontend**

- [x] Add scaffolding to recognize filter predicates https://github.com/sourcegraph/sourcegraph/pull/19641

- [x] Syntax: Parsing and validation routines for tokens https://github.com/sourcegraph/sourcegraph/pull/19785

- [x] Smart autocomplete parentheses for recognized predicates https://github.com/sourcegraph/sourcegraph/pull/20181

- [ ] smart diagnostics for recognizing valid values [WIP](https://github.com/sourcegraph/sourcegraph/tree/rvt/contains-diagnostics)

- [x] Hovers and highlighting https://github.com/sourcegraph/sourcegraph/pull/19894 https://github.com/sourcegraph/sourcegraph/pull/19900 https://github.com/sourcegraph/sourcegraph/pull/19990 https://github.com/sourcegraph/sourcegraph/pull/20192

- [x] Pings/Metrics

**Documentation**

- [x] `select` documentation
- [x] `contains` documentation https://github.com/sourcegraph/sourcegraph/pull/20573

Resolved open question:

- We thought about whether to prefix filter predicates with something like `#` or `+`. No satisfying solution popped out of that discussion, which took place over a couple of weeks. We decided to not add a prefix and just recognize registered predicate names.

**Comment by jensim** [comment]
A may zing!! 😀🥰
I saw the release notes of v3.27.0 `Search functionality via built-in contains predicate: repo:contains(...), repo:contains.file(...), repo:contains.content(...), repo:contains.commit.after(...)`. #18584` and my jaw dropped 😮
It dropped even further when i scrubbed a bit in the RFC and saw `repo:contains(file:README content:foo) file:bar baz`
I've wanted this so bad! The ability to search for repos containing multiple patterns. 

Not sure this is the right forum for this kind of fanboi'ing.. 😊
But I just wanted to say you are doing fantastic stuff!
PS ❤️ `comby`

**Comment by camdencheek** [comment]
@jensim Glad to hear you like it! Rijnard has a bunch of really cool ideas to take this syntax even further with things like `file:has.codeowner(rvantonder)`, so keep your eyes peeled!

**Comment by jensim** [comment]
@camdencheek ooooh 🤩

I just upgraded - and it's the best 😀 Being able to search matching multiple files, it's so simple now! I used to clone every single repo and keep it updated using scripts, and then search using complex xargs+silver-searcher pipes - NO MORE!

Inital reactions:
I love it, I want more! :-D
```
count:all repo:contains(file:pom.xml$ content:>spring-boot-starter<  -content:>junit-vintage-engine<) file:\.(java|kt)$ org.junit.Test
```

> This Search looks for java/kotlin projects running spring-boot-starter, that declare their tests with the old junit4-annotation, but does not declare explicit dependency on the vintage engine dependency. Were I to upgrade the spring-boot-starter version, these projects would not compile, since spring-boot-starter dropped that dependency in one of the newer versions.

This search contains 2 things that i know are not supported (yet)
* negative contains expressions
* multiple content expressions within contains

**Comment by rvantonder** [comment]
@jensim always happy to get feedback and to hear we're solving real pains, please don't hesitate to share (GH issues are fine). It helps me motivate and prioritize search features we work on :-) 

Also, about the junit4 annotation: that's very useful context and example. Negation makes sense to support, I'll see where I can work this in.

**Comment by rvantonder** [comment]
This is done, any follow up work to get separate issue trackers.
